File: 03-SOLID-Principles/src/go/ocp.go
package main

// OCP — Open/Closed Principle
// New discount types = new structs satisfying the interface

type DiscountStrategy interface {
	Calculate(amount float64) float64
}

type RegularDiscount struct{}
func (RegularDiscount) Calculate(amount float64) float64 { return amount * 0.05 }

type PremiumDiscount struct{}
func (PremiumDiscount) Calculate(amount float64) float64 { return amount * 0.10 }

type VIPDiscount struct{}
func (VIPDiscount) Calculate(amount float64) float64 { return amount * 0.20 }

type EmployeeDiscount struct{}
func (EmployeeDiscount) Calculate(amount float64) float64 { return amount * 0.30 }

func CalculateDiscount(s DiscountStrategy, amount float64) float64 {
	return s.Calculate(amount)
}

File: 03-SOLID-Principles/src/rust/ocp.rs
/// OCP — Open/Closed Principle
/// New discount types = new structs implementing the trait

pub trait DiscountStrategy {
    fn calculate(&self, amount: f64) -> f64;
}

pub struct RegularDiscount;
impl DiscountStrategy for RegularDiscount {
    fn calculate(&self, amount: f64) -> f64 { amount * 0.05 }
}

pub struct PremiumDiscount;
impl DiscountStrategy for PremiumDiscount {
    fn calculate(&self, amount: f64) -> f64 { amount * 0.10 }
}

pub struct VIPDiscount;
impl DiscountStrategy for VIPDiscount {
    fn calculate(&self, amount: f64) -> f64 { amount * 0.20 }
}

pub struct EmployeeDiscount;
impl DiscountStrategy for EmployeeDiscount {
    fn calculate(&self, amount: f64) -> f64 { amount * 0.30 }
}

pub fn calculate_discount(strategy: &dyn DiscountStrategy, amount: f64) -> f64 {
    strategy.calculate(amount)
}
